File: ideal_gas.c
#include <stdio.h>   // For printf
#include <stdlib.h>  // For rand, srand, malloc, free
#include <math.h>    // For sqrt, log, cos
#include <time.h>    // For clock(), CLOCKS_PER_SEC

#define L 10.0  // Box size
#define T 1.0   // Temperature scale for velocity distribution
#define DT 0.01 // Time step
#define STEPS 1000  // Number of simulation steps
#define MASS 1.0 // Particle mass

typedef struct {
    double x, y, z;
} Vector3;

// Generate a uniform random number between [0, 1]
double rand_uniform() {
    return rand() / (double)RAND_MAX;
}

// Generate a normally distributed random number using Box-Muller transform
double rand_normal() {
    double u1 = rand_uniform();
    double u2 = rand_uniform();
    return sqrt(-2.0 * log(u1)) * cos(2.0 * M_PI * u2);
}

// Initialize positions uniformly in the box
void initialize_positions(Vector3 *pos, int num_particles) {
    for (int i = 0; i < num_particles; i++) {
        pos[i].x = rand_uniform() * L;
        pos[i].y = rand_uniform() * L;
        pos[i].z = rand_uniform() * L;
    }
}

// Initialize velocities using Maxwell-Boltzmann distribution
void initialize_velocities(Vector3 *vel, int num_particles) {
    for (int i = 0; i < num_particles; i++) {
        double v_mag = sqrt(-2.0 * T * log(rand_uniform()));
        Vector3 v_dir = {rand_normal(), rand_normal(), rand_normal()};
        double norm = sqrt(v_dir.x * v_dir.x + v_dir.y * v_dir.y + v_dir.z * v_dir.z);
        if (norm == 0) norm = 1.0;
        vel[i].x = (v_mag * v_dir.x) / norm;
        vel[i].y = (v_mag * v_dir.y) / norm;
        vel[i].z = (v_mag * v_dir.z) / norm;
    }
}

// Update positions and handle wall collisions
void update_positions(Vector3 *pos, Vector3 *vel, int num_particles) {
    for (int i = 0; i < num_particles; i++) {
        pos[i].x += vel[i].x * DT;
        pos[i].y += vel[i].y * DT;
        pos[i].z += vel[i].z * DT;

        // Wall collision handling
        if (pos[i].x < 0 || pos[i].x > L) {
            vel[i].x *= -1;
            pos[i].x = (pos[i].x < 0) ? 0 : L;
        }
        if (pos[i].y < 0 || pos[i].y > L) {
            vel[i].y *= -1;
            pos[i].y = (pos[i].y < 0) ? 0 : L;
        }
        if (pos[i].z < 0 || pos[i].z > L) {
            vel[i].z *= -1;
            pos[i].z = (pos[i].z < 0) ? 0 : L;
        }
    }
}

// Compute temperature, net velocity, and pressure
void compute_properties(Vector3 *vel, int num_particles, double *temperature, double *net_velocity, double *pressure) {
    double kinetic_energy = 0.0;
    Vector3 momentum = {0.0, 0.0, 0.0};
    
    for (int i = 0; i < num_particles; i++) {
        double v_sq = vel[i].x * vel[i].x + vel[i].y * vel[i].y + vel[i].z * vel[i].z;
        kinetic_energy += 0.5 * MASS * v_sq;
        momentum.x += vel[i].x;
        momentum.y += vel[i].y;
        momentum.z += vel[i].z;
    }

    *temperature = (2.0 / 3.0) * (kinetic_energy / num_particles);
    *net_velocity = sqrt(momentum.x * momentum.x + momentum.y * momentum.y + momentum.z * momentum.z) / num_particles;
    *pressure = (2.0 * kinetic_energy) / (3.0 * L * L * L);
}

int main(int argc, char *argv[]) {
    clock_t start_time, end_time;
    start_time = clock();  // Start timing execution

    srand(time(NULL)); // Seed random number generator

    if (argc != 2) {fprintf(stderr, "Usage: %s <number_of_particles>\n", argv[0]);
        return 1;
    }

       // Convert input argument to integer
    int N = atoi(argv[1]);
    if (N <= 0) {
        fprintf(stderr, "Error: Number of particles must be a positive integer.\n");
        return 1;
    }


    Vector3 *pos = (Vector3 *)malloc(N * sizeof(Vector3));
    Vector3 *vel = (Vector3 *)malloc(N * sizeof(Vector3));

    initialize_positions(pos, N);
    initialize_velocities(vel, N);

    for (int step = 0; step < STEPS; step++) {
        update_positions(pos, vel, N);
    }

    double temperature, net_velocity, pressure;
    compute_properties(vel, N, &temperature, &net_velocity, &pressure);

    printf("System Properties:\n");
    printf("Temperature: %.2f\n", temperature);
    printf("Net Velocity: %.2f\n", net_velocity);
    printf("Pressure: %.2f\n", pressure);

    free(pos);
    free(vel);

    end_time = clock();  // Stop timing execution
    double execution_time = (double)(end_time - start_time) / CLOCKS_PER_SEC;
    printf("Execution Time: %.6f seconds\n", execution_time);

    return 0;
}

File: ideal_gas.py
import numpy as np
import sys
import time

# Constants
L = 10.0  # Box size
T = 1.0  # Arbitrary temperature scale for velocity distribution
dt = 0.01  # Time step
steps = 1000  # Number of simulation steps

# Check if the number of particles is passed as a command-line argument
if len(sys.argv) != 2:
    print("Usage: python ideal_gas.py <num_particles>")
    sys.exit(1)

N = int(sys.argv[1])  # Number of particles

# Initialize positions (uniformly distributed in the box)
pos = np.random.uniform(0, L, size=(N, 3))

# Initialize velocities from Maxwell-Boltzmann distribution manually
v_mag = np.sqrt(-2 * T * np.log(np.random.uniform(size=N)))  # Maxwell-Boltzmann speed
v_dir = np.random.normal(size=(N, 3))  # Random directions
v_dir /= np.linalg.norm(v_dir, axis=1, keepdims=True)  # Normalize
vel = v_mag[:, np.newaxis] * v_dir  # Assign velocities

# Track wall collisions for pressure calculation
wall_collisions = 0

def update_positions(pos, vel, L, dt):
    """Update particle positions and handle wall collisions."""
    global wall_collisions
    pos += vel * dt  # Move particles

    for i in range(len(pos)):
        for j in range(3):  # Check each coordinate (x, y, z)
            if pos[i, j] < 0 or pos[i, j] > L:
                vel[i, j] *= -1  # Reverse velocity
                pos[i, j] = max(0, min(L, pos[i, j]))  # Keep inside box
                wall_collisions += 1  # Count collision

    return pos, vel

def net_velocity(vel):
    """Calculate the net velocity vector (should be close to zero)."""
    return np.sum(vel, axis=0)  # Sum of all velocity components

def check_energy_conservation(vel_initial, vel_final):
    """Check if total kinetic energy is conserved."""
    KE_initial = np.sum(0.5 * np.sum(vel_initial**2, axis=1))  # KE before
    KE_final = np.sum(0.5 * np.sum(vel_final**2, axis=1))  # KE after
    return np.isclose(KE_initial, KE_final)

def calculate_temperature(vel):
    """Estimate temperature from mean velocity squared."""
    v_mean_sq = np.mean(np.sum(vel**2, axis=1))
    return v_mean_sq / 3  # Approximate T

def estimate_pressure(wall_collisions, V, N, T):
    """Estimate pressure using the ideal gas approximation."""
    return (N * T) / V  # Using P ≈ NkT/V (simplified)

# Store initial velocities for energy check
vel_initial = vel.copy()

# Start time measurement
start_time = time.time()

# Run simulation
for _ in range(steps):
    pos, vel = update_positions(pos, vel, L, dt)

# Perform testing
net_vel = net_velocity(vel)
energy_conserved = check_energy_conservation(vel_initial, vel)
estimated_T = calculate_temperature(vel)
estimated_P = estimate_pressure(wall_collisions, L**3, N, estimated_T)

# End time measurement
end_time = time.time()

# Display results
print(f"Net velocity vector: {net_vel}")
print(f"Energy conserved: {energy_conserved}")
print(f"Estimated Temperature: {estimated_T:.3f}")
print(f"Estimated Pressure: {estimated_P:.3f}")
print(f"Execution Time: {end_time - start_time:.6f} seconds")
